narwhals/dtypes.py

```diff
@@ -805,16 +828,24 @@ def __eq__(self: Self, other: DType | type[DType]) -> bool:  # type: ignore[over
         if type(other) is type and issubclass(other, self.__class__):
             return True
         elif isinstance(other, self.__class__):
-            return self.inner == other.inner
+            if self.shape != other.shape:
+                return False
+            else:
+                return self.inner == other.inner
         else:
             return False
 
     def __hash__(self: Self) -> int:
-        return hash((self.__class__, self.inner, self.width))
+        return hash((self.__class__, self.inner, self.shape))
+
+    def __repr__(self) -> str:
+        # Get leaf type
+        dtype = self.inner
+        while isinstance(dtype, Array):
+            dtype = dtype.inner
```

Comment:
can we put a limit on the depth of this (just to avoid some infinite loop in some unexpected scenario)?
same with the other place where there's the while loop

Reply:
This in particular is the same as polars does it: https://github.com/pola-rs/polars/blob/faad12f7277751006e3faebf0fffb1f6bf9aa7e7/py-polars/polars/datatypes/classes.py#L905-L912

and the `__init__` is also recursive:

https://github.com/narwhals-dev/narwhals/blob/981f87c30a7d9284f5730ddd36c5162e111de194/narwhals/dtypes.py#L808-L810

In principle, max depth should be `len(shape)`. Is that a good depth to check? WDYT?

Reply:
i was thinking something like

https://github.com/psf/black/blob/5f2370170819d282ec14dcda70f963d7574271e2/src/black/handle_ipynb_magics.py#L217-L226

Reply:
I remember the trick, yet I don't feel super comfortable with introducing a fixed number of depth to check for.

Since we know already that the max depth should be `len(shape)`, I think that would be the best limit to introduce, if any.

In practice, if someone generate a datatype to pass which ends up in an infinite recursion, maybe it's good to raise with sure error instead of an `AssertionError`. I am thinking out loud here though

Reply:
Sure, shall we use len shape? 

In theory this shouldn't happen, just trying to guard against implementation oversights